zebra-rpc/src/queue.rs

```diff
@@ -0,0 +1,297 @@
+            }
+        }
+    }
+}
+
+impl Runner {
+    /// Access the sender field of the runner.
+    pub fn sender(&self) -> Sender<Option<UnminedTx>> {
+        self.sender.clone()
+    }
+
+    /// Access the mutable queue.
+    pub fn queue(&self) -> Arc<Mutex<Queue>> {
+        self.queue.clone()
+    }
+
+    /// Get the queue transactions as a `HashSet` of unmined ids.
+    fn transactions_as_hash_set(&self) -> HashSet<UnminedTxId> {
+        let transactions = self
+            .queue
+            .lock()
+            .expect("queue mutex should be unpoisoned")
+            .transactions();
+        transactions.iter().map(|t| *t.0).collect()
+    }
+
+    /// Get the queue transactions as a `Vec` of transactions.
+    fn transactions_as_vec(&self) -> Vec<Arc<Transaction>> {
+        let transactions = self
+            .queue
+            .lock()
+            .expect("queue mutex should be unpoisoned")
+            .transactions();
+        transactions.iter().map(|t| t.1 .0.clone()).collect()
+    }
+
+    /// Retry sending to memempool if needed.
+    pub async fn run<Mempool, State, Tip>(self, mempool: Mempool, state: State, _tip: Tip)
+    where
+        Mempool: Service<Request, Response = Response, Error = BoxError> + Clone + 'static,
+        State: Service<ReadRequest, Response = ReadResponse, Error = zebra_state::BoxError>
+            + Clone
+            + Send
+            + Sync
+            + 'static,
+        Tip: ChainTip + Clone + Send + Sync + 'static,
+    {
+        // TODO: this should be an argument
+        let network = zebra_chain::parameters::Network::Mainnet;
+
+        // TODO: Use tip.best_tip_height()
+        let tip_height = zebra_chain::block::Height(1);
+
+        // get spacing between blocks
+        let spacing = NetworkUpgrade::target_spacing_for_height(network, tip_height);
+
+        loop {
+            // sleep until the next block
+            tokio::time::sleep(spacing.to_std().unwrap()).await;
+
+            // first, remove what is expired
+            self.remove_expired(spacing);
+
+            // remove if any of the queued transactions is now in the mempool
+            let in_mempool =
+                Self::check_mempool(mempool.clone(), self.transactions_as_hash_set()).await;
+            self.remove_committed(in_mempool);
+
+            // remove if any of the queued transactions is now in the state
+            let in_state = Self::check_state(state.clone(), self.transactions_as_hash_set()).await;
+            self.remove_committed(in_state);
+
+            // retry what is left in the queue
+            let _retried = Self::retry(mempool.clone(), self.transactions_as_vec()).await;
+        }
+    }
+
+    /// Remove transactions that are expired according to number of blocks and current spacing between blocks.
+    fn remove_expired(&self, spacing: Duration) {
+        let duration_to_expire =
+            Duration::seconds(NUMBER_OF_BLOCKS_TO_EXPIRE * spacing.num_seconds());
```

Comment:
If you want to make sure you re-submit each transaction `NUMBER_OF_BLOCKS_TO_EXPIRE` times, add some extra time here.

Each loop also takes some time to run, on top of the `sleep` time.